File: internal/models/anthropic.go
package models

import (
	"encoding/json"
	"fmt"
)

// MessagesRequest is the Anthropic /v1/messages inbound payload.
type MessagesRequest struct {
	Model     string           `json:"model"`
	System    FlexContent      `json:"system,omitempty"`
	Messages  []AnthropicTurn  `json:"messages"`
	Stream    bool             `json:"stream"`
	MaxTokens int              `json:"max_tokens,omitempty"`
}

// AnthropicTurn is one role/content pair in the messages array.
type AnthropicTurn struct {
	Role    string      `json:"role"`
	Content FlexContent `json:"content"`
}

// AnthropicDeltaEvent is a content_block_delta SSE data payload.
type AnthropicDeltaEvent struct {
	Type  string `json:"type"`
	Index int    `json:"index"`
	Delta struct {
		Type string `json:"type"`
		Text string `json:"text"`
	} `json:"delta"`
}

// ExtractPromptState returns system text and the latest user message for cache keying.
func (r *MessagesRequest) ExtractPromptState() (systemPrompt, latestUser string) {
	systemPrompt = r.System.Text()
	for _, msg := range r.Messages {
		if msg.Role == "user" {
			latestUser = msg.Content.Text()
		}
	}
	return systemPrompt, latestUser
}

// Clone returns a deep copy suitable for middleware mutation.
func (r *MessagesRequest) Clone() *MessagesRequest {
	out := *r
	out.Messages = make([]AnthropicTurn, len(r.Messages))
	copy(out.Messages, r.Messages)
	return &out
}

// Marshal serializes the request to JSON bytes.
func (r *MessagesRequest) Marshal() ([]byte, error) {
	return json.Marshal(r)
}

// ParseMessagesRequest decodes an Anthropic messages body.
func ParseMessagesRequest(body []byte) (*MessagesRequest, error) {
	var req MessagesRequest
	if err := json.Unmarshal(body, &req); err != nil {
		return nil, fmt.Errorf("parse messages request: %w", err)
	}
	return &req, nil
}

File: rust/korto-proxy/src/models/anthropic.rs
//! Anthropic request models — mirrors `internal/models/anthropic.go`.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::openai::content_text;

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct MessagesRequest {
    pub model: String,
    #[serde(default)]
    pub system: Value,
    pub messages: Vec<AnthropicTurn>,
    #[serde(default)]
    pub stream: bool,
    #[serde(default)]
    pub max_tokens: u32,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct AnthropicTurn {
    pub role: String,
    pub content: Value,
}

impl MessagesRequest {
    pub fn extract_prompt_state(&self) -> (String, String) {
        let system_prompt = content_text(&self.system);
        let mut latest_user = String::new();
        for msg in &self.messages {
            if msg.role == "user" {
                latest_user = content_text(&msg.content);
            }
        }
        (system_prompt, latest_user)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn extracts_prompt_state() {
        let req: MessagesRequest = serde_json::from_value(json!({
            "model": "claude-3-5-sonnet-20241022",
            "max_tokens": 64,
            "stream": true,
            "system": "sys",
            "messages": [{"role": "user", "content": "ping"}]
        }))
        .unwrap();
        assert_eq!(req.extract_prompt_state(), ("sys".into(), "ping".into()));
    }
}
